ustack: cache the base addresses

alban opened the pull request:
Fixes: https://github.com/inspektor-gadget/inspektor-gadget/issues/4928

Due to Address Space Layout Randomization (ASLR), most processes have a random base address chosen at execve-time.

Before this patch, IG parsed `/proc/pid/maps` to find the base address at each stack trace resolution. This patch adds a cache to minimize the overhead.

The cache is not indexed by pid because:
- The base address changes with execve but the pid does not change at that time
- Several pids can share the same memory space with CLONE_VM (e.g. threads)

Instead, I use the mm_struct address to distinguish the programs sharing the same base address. There are however a possibility where the `mm_struct` address is reused: if the `mm_struct` is freed and re-allocated to the same address. To avoid this scenario, I add `mm->start_code` and `mm->start_stack` in the hash. A new program reusing the same mm_struct address will have a different random `start_code` and `start_stack`.

To reduce the size of events sent to the ebpf ring buffer, we send two u32 hashes:

```c
	// Identify the executable file.
	// Changes when the executable file changes.
	// Shared between processes using the same executable.
	// Can be used as index for cache of symbol tables.
	// FNV-1a hash of tuple (exe_inode, mtime_sec, mtime_nsec)
	__u32 exe_hash;

	// Identify the base address.
	// Changes whenever the base address changes at execve.
	// Not shared between processes because of Address Space Layout Randomization.
	// FNV-1a hash of tuple (mm_addr, start_code, start_stack)
	__u32 base_addr_hash;
```

I chose the hash algorithm FNV-1a because it is fast and easy to implement in ebpf. It is not cryptographically secure but the 3 components of the tuple `(mm_addr, start_code, start_stack)` are chosen by the kernel.

Note that the base address cannot reliably be found from ebpf. Even though `mm_struct` has fields like `start_code` and `start_data`, that does not directly give the base address. The information could be found in kernel space by iterating over the VMAs but this cannot be done in ebpf because that would be an unbounded loop.

Example:
```
$ sudo cat /proc/1853244/maps
55f86c7ec000-55f86c7ed000 r--p 00000000 08:16 23894993                   /usr/local/bin/python3.13
55f86c7ed000-55f86c7ee000 r-xp 00001000 08:16 23894993                   /usr/local/bin/python3.13
55f86c7ee000-55f86c7ef000 r--p 00002000 08:16 23894993                   /usr/local/bin/python3.13
55f86c7ef000-55f86c7f0000 r--p 00002000 08:16 23894993                   /usr/local/bin/python3.13
55f86c7f0000-55f86c7f1000 rw-p 00003000 08:16 23894993                   /usr/local/bin/python3.13
...
55f86d141000-55f86d704000 rw-p 00000000 00:00 0                          [heap]
...
7ffd65563000-7ffd65584000 rw-p 00000000 00:00 0                          [stack]
```

`mm_struct` fields (found by [modifying the snapshot_process gadget](https://github.com/alban/inspektor-gadget/commit/9d65a3c80f1d86a16b2c805c03193b267966183a)):

| Field Name       | Address (Hexadecimal) |
| :---             | :---                  |
| **start_code**   | 55f86c7ed000          |
| **end_code**     | 55f86c7ed155          |
| **start_data**   | 55f86c7efdb0          |
| **end_data**     | 55f86c7f0018          |
| **start_brk**    | 55f86d141000          |
| **start_stack**  | 7ffd65583590          |

The base address is `55f86c7ec000` and the `start_code` is `55f86c7ed000` (8KiB below). While it is often 8KiB in practice, we cannot rely on that. It is 8KiB here because the first LOAD segment has a size of 8KiB (after rounding up on the page size). It typically contains metadata like the ELF header, the program headers and the interpreter path ("/lib64/ld-linux-x86-64.so.2"). Other programs might be compiled differently.

```
$ sudo readelf -lW /usr/local/bin/python3.13
Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  PHDR           0x000040 0x0000000000000040 0x0000000000000040 0x000310 0x000310 R   0x8
  INTERP         0x000394 0x0000000000000394 0x0000000000000394 0x00001c 0x00001c R   0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  LOAD           0x000000 0x0000000000000000 0x0000000000000000 0x000760 0x000760 R   0x1000
  LOAD           0x001000 0x0000000000001000 0x0000000000001000 0x000155 0x000155 R E 0x1000
  LOAD           0x002000 0x0000000000002000 0x0000000000002000 0x0000f0 0x0000f0 R   0x1000
  LOAD           0x002db0 0x0000000000003db0 0x0000000000003db0 0x000268 0x000270 RW  0x1000
...
```

## How to use

See testing below.

## Testing done

I used the following commands to test:
```
sudo ig run ghcr.io/inspektor-gadget/gadget/trace_capabilities:alban_ustack_hash --verify-image=false --collect-ustack --comm=chroot1 --host -v
sudo ig run ghcr.io/inspektor-gadget/gadget/trace_capabilities:alban_ustack_hash --verify-image=false --collect-ustack --comm=chroot2 --host -v
```

The workload is a process that generates 2 events and then calls execve itself several time but without changing the executable file:

<details>
<summary>chroot.c</summary>

```c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>

__attribute__((noinline)) void level3() {
    int ret = chroot("/");
    printf("chroot: %d\n", ret);
}
__attribute__((noinline)) void level2() {
    level3();
}
__attribute__((noinline)) void level1() {
    level2();
}

int main(int argc, char *argv[]) {
    level1();
    sleep(1);
    level1();
    sleep(1);
    
    if (argc > 1) {
        execvp(argv[1], &argv[1]);
        perror("execvp failed");
        return 1;
    }
    
    return 0;
}
```

</details>

<details>
<summary>Makefile</summary>

```Makefile
all:
	gcc -Wall            -o chroot1 chroot.c
	gcc -Wall -fPIE -pie -o chroot2 chroot.c
```

</details>

```
$ ./chroot1 ./chroot1 ./chroot1 ./chroot1 /bin/sleep 1
```
The runtime base address is recalculated only after a execve, and taken from the cache otherwise.

<details>
<summary>IG output</summary>

```
DEBU[0079] symbol table for "/chroot1" (pid 2719252) loaded: 9 symbols (total: 2 symbol tables with 18 symbols) 
DEBU[0079] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from /proc/2719252/maps) 
DEBU[0079] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0080] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from cache) 
DEBU[0080] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0081] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from /proc/2719252/maps) 
DEBU[0081] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0082] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from cache) 
DEBU[0082] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0083] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from /proc/2719252/maps) 
DEBU[0083] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0084] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from cache) 
DEBU[0084] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0085] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from /proc/2719252/maps) 
DEBU[0085] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0086] getRuntimeBaseAddr: pid 2719252 (/chroot1) runtime base address: 0x400000 (from cache) 
DEBU[0086] resolving 7 stack frames with symbol table (pid 2719252, PIE=false): runtime base address 0x400000, elf base address 0x400000, bias 0 
                                                                                   chroot1             2719252    2719252 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
```

</details>

It also works with PIE executables:
```
$ ./chroot2 ./chroot2 ./chroot2 ./chroot2 /bin/sleep 1
```

<details>
<summary>IG output</summary>

```
DEBU[0006] symbol table for "/chroot2" (pid 2723119) loaded: 8 symbols (total: 1 symbol tables with 8 symbols) 
DEBU[0006] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x55f9a8d6d000 (from /proc/2723119/maps) 
DEBU[0006] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x55f9a8d6d000, elf base address 0x0, bias -94530767867904 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0007] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x55f9a8d6d000 (from cache) 
DEBU[0007] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x55f9a8d6d000, elf base address 0x0, bias -94530767867904 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0008] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x563e296dd000 (from /proc/2723119/maps) 
DEBU[0008] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x563e296dd000, elf base address 0x0, bias -94824983023616 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0009] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x563e296dd000 (from cache) 
DEBU[0009] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x563e296dd000, elf base address 0x0, bias -94824983023616 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0010] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x559f3f4a2000 (from /proc/2723119/maps) 
DEBU[0010] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x559f3f4a2000, elf base address 0x0, bias -94142449983488 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0011] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x559f3f4a2000 (from cache) 
DEBU[0011] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x559f3f4a2000, elf base address 0x0, bias -94142449983488 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0012] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x5565362be000 (from /proc/2723119/maps) 
DEBU[0012] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x5565362be000, elf base address 0x0, bias -93893188902912 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
DEBU[0013] getRuntimeBaseAddr: pid 2723119 (/chroot2) runtime base address: 0x5565362be000 (from cache) 
DEBU[0013] resolving 7 stack frames with symbol table (pid 2723119, PIE=true): runtime base address 0x5565362be000, elf base address 0x0, bias -93893188902912 
                                                                                   chroot2             2723119    2723119 false 1           CAP_SYS_CHROOT                              SYS_CHROOT                                            
```

</details>

alban commented:
I rebased this PR on top of #4932 so that I can test it with Go programs.

mauriciovasquezbernal commented on pkg/symbolizer/symtab_linux.go:
```diff
@@ -123,13 +146,13 @@ func (s *Symbolizer) newSymbolTableFromPid(pid uint32, expectedExeKey exeKey) (*
 		return nil, errors.New("getting syscall.Stat_t failed")
 	}
 	ino := stat.Ino
```
nit: this variable can be removed and use stat.Ino directly on the error message.

mauriciovasquezbernal commented on pkg/symbolizer/symtab_linux.go:
```diff
@@ -251,7 +277,13 @@ func (e *symbolTable) lookupByAddr(address uint64) string {
 // getRuntimeBaseAddr gets the runtime base address of the main executable from /proc/pid/maps
-func getRuntimeBaseAddr(pid uint32) (uint64, error) {
+func getRuntimeBaseAddr(task Task, table *symbolTable, pid uint32) (uint64, error) {
+	if runtimeBaseAddr := table.runtimeBaseAddrCache[task.BaseAddrHash]; runtimeBaseAddr != 0 {
```
I suppose runtimeBaseAddr can never be zero so we don't need to distinguish between not found and 0, right?

mauriciovasquezbernal reviewed (approved):
Some minor nits but LGTM from code inspection.

mauriciovasquezbernal commented on include/gadget/fnv1a.h:
```diff
@@ -0,0 +1,55 @@
+		}                                                              \
+	}
+
+// 32-bit hash functions
+DEFINE_FNV_UPDATE(fnv_32a_update_u64, u32, u64, sizeof(u64), FNV1_32_PRIME)
+DEFINE_FNV_UPDATE(fnv_32a_update_u32, u32, u32, sizeof(u32), FNV1_32_PRIME)
+DEFINE_FNV_UPDATE(fnv_32a_update_u16, u32, u16, sizeof(u16), FNV1_32_PRIME)
+DEFINE_FNV_UPDATE(fnv_32a_update_u8, u32, u8, sizeof(u8), FNV1_32_PRIME)
+
+// 64-bit hash functions
+DEFINE_FNV_UPDATE(fnv_64a_update_u64, u64, u64, sizeof(u64), FNV1_64_PRIME)
+DEFINE_FNV_UPDATE(fnv_64a_update_u32, u64, u32, sizeof(u32), FNV1_64_PRIME)
+DEFINE_FNV_UPDATE(fnv_64a_update_u16, u64, u16, sizeof(u16), FNV1_64_PRIME)
+DEFINE_FNV_UPDATE(fnv_64a_update_u8, u64, u8, sizeof(u8), FNV1_64_PRIME)
```
Up to you, making them more private to avoid conflicts is good, but `fnv_32a_update_u64` is already too specific and a collision is unlikely.

alban commented:
For the record, I updated this PR with some differences compared to the initial implementation:
- I added the device major/minor to ensure inode uniqueness among filesystems
- I reverted the change to make a hash from the tuple (major, minor, inode, mtime), so now the complete tuple is sent to userspace and used as a key to the SymbolTable cache. This is to avoid any risks of hash collision.
- I use a different base address cache for each process group, so two tasks cannot share the same cache unless they are threads of the same process.
